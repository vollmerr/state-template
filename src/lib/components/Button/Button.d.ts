import * as React from 'react';

import * as Icon from '../Icon/Icon';

export interface ButtonProps {
  /** Use style variant */
  variant?:  ''
    | 'default'
    | 'primary'
    | 'secondary'
    | 'highlight'
    | 'standout';

  /** Style as active */
  active?: boolean;

  /** Base button class to apply */
  baseClass?: string;

  /** Icon to render */
  iconProps: Icon.IconProps;

  /** HTML tag to render as */
  tag?: React.ReactNode;

  /** Content to render, mutually exlusive with `children` */
  text?: React.ReactNode;

  /** Button type */
  type?: ''
    | 'button'
    | 'submit'
    | 'reset';

  /** Style class name to attach to button */
  className?: string;

  /** HTML tag to render as */
  tag?: string | React.ReactNode;

  /** Content to render */
  children?: React.ReactNode;

  /** Icon to render */
  iconProps: Icon.IconProps;
}

declare class Button extends React.Component<ButtonProps, {}> {}

export default Button;
