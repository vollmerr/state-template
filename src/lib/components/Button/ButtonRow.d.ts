import * as React from 'react';

export interface ButtonRowProps {
  /** Style class name to attach to button */
  className?: string;

  /** Content to render */
  children?: React.ReactNode;
}

declare class ButtonRow extends React.Component<ButtonRowProps, {}> {}

export default ButtonRow;
