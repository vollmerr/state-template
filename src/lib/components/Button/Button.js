import React from 'react';
import classNames from 'classnames';
import T from 'prop-types';

import Icon from '../Icon';

/**
 * Buttons come in various styles and inherit
 * all properties of a standard html button.
 */
const Button = (props) => {
  const {
    variant,
    active,
    text,
    className,
    tag: Tag,
    children,
    iconProps,
    ...rest
  } = props;

  const btnClass = classNames([
    'btn',
    className,
    { active, [`btn-${variant}`]: variant },
  ]);

  return (
    <Tag data-test={'button'} className={btnClass} {...rest}>
      {iconProps && <><Icon {...iconProps} />{' '}</>}
      {text || children}
    </Tag>
  );
};

Button.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'default',
    'primary',
    'secondary',
    'highlight',
    'standout',
  ]),

  /** Style as active */
  active: T.bool,

  /** Content to render, mutually exclusive with `children` */
  text: T.node,

  /** Button type */
  type: T.oneOf([
    '',
    'button',
    'submit',
    'reset',
  ]),

  /** Style class name to attach to button */
  className: T.string,

  /** HTML tag to render as */
  tag: T.oneOfType([T.func, T.string]),

  /** Content to render */
  children: T.node,

  /** Icon to render */
  iconProps: T.shape({ ...Icon.propTypes }),
};

Button.defaultProps = {
  variant: '',
  active: false,
  text: null,
  type: 'button',
  className: '',
  tag: 'button',
  children: null,
  iconProps: null,
};

export default Button;
