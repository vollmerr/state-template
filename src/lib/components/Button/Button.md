```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'button'} />
  <Button text={'default button'} variant={'default'} />
  <Button text={'primary button'} variant={'primary'} />
  <Button text={'secondary button'} variant={'secondary'} />
  <Button text={'highlight button'} variant={'highlight'} />
  <Button text={'standout button'} variant={'standout'} />
</ButtonRow>
```

### active

```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'button'} />
  <Button text={'default button'} variant={'default'} active />
  <Button text={'primary button'} variant={'primary'} active />
  <Button text={'secondary button'} variant={'secondary'} active />
  <Button text={'highlight button'} variant={'highlight'} active />
  <Button text={'standout button'} variant={'standout'} active />
</ButtonRow>
```

### disabled

```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'button'} disabled />
  <Button text={'default button'} variant={'default'} disabled />
  <Button text={'primary button'} variant={'primary'} disabled />
  <Button text={'secondary button'} variant={'secondary'} disabled />
  <Button text={'highlight button'} variant={'highlight'} disabled />
  <Button text={'standout button'} variant={'standout'} disabled />
</ButtonRow>
```

### with icon

```jsx
import { ButtonRow } from 'state-template';

const iconProps = {
  name: 'plus-mark',
};

<ButtonRow>
  <Button text={'button'} iconProps={iconProps} />
  <Button text={'default button'} variant={'default'} iconProps={iconProps} />
  <Button text={'primary button'} variant={'primary'} iconProps={iconProps} />
  <Button text={'secondary button'} variant={'secondary'} iconProps={iconProps} />
  <Button text={'highlight button'} variant={'highlight'} iconProps={iconProps} />
  <Button text={'standout button'} variant={'standout'} iconProps={iconProps} />
</ButtonRow>
```

### passed props
Other properties get passed down to the button component, such as onChange handlers, id, etc.

```jsx
import { ButtonRow } from 'state-template';

<ButtonRow>
  <Button text={'click me'} onClick={() => alert('You Clicked Me!!')} variant={'default'} />
  <Button text={'other props'} id={'ButtonId'} style={{ fontSize: '20px' }} variant={'primary'} />
</ButtonRow>
```
