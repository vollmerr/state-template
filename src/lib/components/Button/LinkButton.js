import React from 'react';
import T from 'prop-types';
import { Link } from 'react-router-dom';

import Button from '.';

/**
 * A `Button` specialized for links. Buttons with a `to`
 * are handled by react-router-dom's internal routing.
 * See [Button](#button) for full set of available props.
 */
const LinkButton = (props) => {
  const {
    to,
    href,
    ...rest
  } = props;

  let btnProps;
  if (href) {
    // is external link
    btnProps = {
      ...rest,
      href,
      target: '_blank',
      tag: 'a',
    };
  } else {
    // is internal link
    btnProps = {
      ...rest,
      to,
      tag: Link,
    };
  }

  // clear type to render as anchor (css issues..)
  btnProps.type = '';

  return <Button {...btnProps} />;
};

LinkButton.propTypes = {
  /** Path for internal links */
  to: T.string,
  /** Path for external links (open in new tab) */
  href: T.string,
};

LinkButton.defaultProps = {
  to: '',
  href: '',
};

export default LinkButton;
