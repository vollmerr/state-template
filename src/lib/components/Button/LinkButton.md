### internal

```jsx
<ButtonRow>
  <LinkButton text={'link'} to={'/'} />
  <LinkButton text={'default link'} to={'/'} variant={'default'} />
  <LinkButton text={'primary link'} to={'/'} variant={'primary'} />
  <LinkButton text={'secondary link'} to={'/'} variant={'secondary'} />
  <LinkButton text={'highlight link'} to={'/'} variant={'highlight'} />
  <LinkButton text={'standout link'} to={'/'} variant={'standout'} />
</ButtonRow>
```

### external
External links open in a new tab.

```jsx
<ButtonRow>
  <LinkButton text={'link'} href={'http://www.google.com'} />
  <LinkButton text={'default link'} href={'http://www.google.com'} variant={'default'} />
  <LinkButton text={'primary link'} href={'http://www.google.com'} variant={'primary'} />
  <LinkButton text={'secondary link'} href={'http://www.google.com'} variant={'secondary'} />
  <LinkButton text={'highlight link'} href={'http://www.google.com'} variant={'highlight'} />
  <LinkButton text={'standout link'} href={'http://www.google.com'} variant={'standout'} />
</ButtonRow>
```
