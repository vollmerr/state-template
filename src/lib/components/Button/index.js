import React from 'react';
import classNames from 'classnames';
import T from 'prop-types';

/**
 * Buttons come in various styles and inherit
 * all properties of a standard html button.
 */
const Button = (props) => {
  const {
    variant,
    isActive,
    text,
    className,
    tag: Tag,
    ...rest
  } = props;

  const btnClass = classNames([
    'btn',
    className,
    { active: isActive, [`btn-${variant}`]: variant },
  ]);

  return <Tag className={btnClass} {...rest}>{text}</Tag>;
};

Button.propTypes = {
  /** Use style variant */
  variant: T.oneOf([
    '',
    'default',
    'primary',
    'secondary',
    'highlight',
    'standout',
  ]),
  /** Text to display */
  text: T.string.isRequired,
  /** Button type */
  type: T.oneOf([
    'button',
    'submit',
    'reset',
  ]),
  /** Style class name to attach to button */
  className: T.string,
  /** HTML tag to use */
  tag: T.oneOfType([T.func, T.string]),
};

Button.defaultProps = {
  variant: '',
  type: 'button',
  className: '',
  tag: 'button',
};

export default Button;
