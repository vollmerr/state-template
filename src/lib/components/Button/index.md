### default

```jsx
<div className={'btn-row'}>
  <Button text={'default button'} />
  <Button text={'primary button'} primary />
  <Button text={'secondary button'} secondary />
  <Button text={'highlight button'} highlight />
  <Button text={'standout button'} standout />
</div>
```

### active

```jsx
<div className={'btn-row'}>
  <Button text={'default button'} isActive />
  <Button text={'primary button'} primary isActive />
  <Button text={'secondary button'} secondary isActive />
  <Button text={'highlight button'} highlight isActive />
  <Button text={'standout button'} standout isActive />
</div>
```

### disabled

```jsx
<div className={'btn-row'}>
  <Button text={'default button'} disabled />
  <Button text={'primary button'} primary disabled />
  <Button text={'secondary button'} secondary disabled />
  <Button text={'highlight button'} highlight disabled />
  <Button text={'standout button'} standout disabled />
</div>
```

### passed props
Other properties get passed down to the button component, such as onChange handlers, id, etc.

```jsx
<div className={'btn-row'}>
  <Button text={'click me'} onClick={() => alert('You Clicked Me!!')} />
  <Button text={'other props'} id={'ButtonId'} style={{ fontSize: '20px' }} />
</div>
```
