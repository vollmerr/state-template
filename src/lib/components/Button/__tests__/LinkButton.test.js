import React from 'react';
import { shallow } from 'enzyme';
import { Link } from 'react-router-dom';

import { LinkButton as fromLib } from '../../..';
import LinkButton from '../LinkButton';
import Button from '..';

const props = {
  text: 'test text',
  href: 'http://www.google.com',
};

let wrapper;
describe('LinkButton', () => {
  beforeEach(() => {
    wrapper = shallow(<LinkButton {...props} />);
  });

  it('should be exported as part of the library', () => {
    expect(fromLib).toEqual(LinkButton);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('external links', () => {
    it('should use an `a` tag', () => {
      const component = wrapper.find(Button).prop('component');
      expect(component).toBe('a');
    });

    it('should open in a new tab', () => {
      const target = wrapper.find(Button).prop('target');
      expect(target).toBe('_blank');
    });
  });

  describe('internal links', () => {
    it('should use a `Link` tag', () => {
      wrapper.setProps({ to: '/testPath', href: undefined });
      const component = wrapper.find(Button).prop('component');
      expect(component).toBe(Link);
    });
  });

  it('should pass additional props', () => {
    wrapper.setProps({ id: 'testId', type: 'submit', variant: 'primary' });
    const button = wrapper.find(Button);
    expect(button.prop('id')).toBe('testId');
    expect(button.prop('type')).toBe('submit');
    expect(button.prop('variant')).toBe('primary');
  });
});
