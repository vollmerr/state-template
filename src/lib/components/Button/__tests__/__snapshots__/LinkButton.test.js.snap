// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkButton should render correctly 1`] = `
<Button
  className=""
  href="http://www.google.com"
  tag="a"
  target="_blank"
  text="test text"
  type="button"
  variant=""
/>
`;
